healthmanager/blockchain_health_manager.go

```diff
@@ -0,0 +1,240 @@
+func (b *BlockchainHealthManager) RegisterProvidersHealthManager(ctx context.Context, phm *ProvidersHealthManager) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	chainID := phm.ChainID()
+
+	// Check if a provider for the given chainID is already registered and remove it
+	if _, exists := b.providers[chainID]; exists {
+		// Cancel the existing context
+		if cancel, cancelExists := b.cancelFuncs[chainID]; cancelExists {
+			cancel()
+		}
+		// Remove the old registration
+		delete(b.providers, chainID)
+		delete(b.cancelFuncs, chainID)
+	}
+
+	// Proceed with the registration
+	b.providers[chainID] = phm
+
+	// Create a new context for the provider
+	providerCtx, cancel := context.WithCancel(ctx)
+	b.cancelFuncs[chainID] = cancel
+
+	statusCh := phm.Subscribe()
+	b.wg.Add(1)
+	go func(phm *ProvidersHealthManager, statusCh chan struct{}, providerCtx context.Context) {
+		defer func() {
+			phm.Unsubscribe(statusCh)
+			b.wg.Done()
+		}()
+		for {
+			select {
+			case <-statusCh:
+				// When the provider updates its status, check the statuses of all providers
+				b.aggregateAndUpdateStatus(providerCtx)
+			case <-providerCtx.Done():
+				// Stop processing when the context is cancelled
+				return
+			}
+		}
+	}(phm, statusCh, providerCtx)
+
+	return nil
+}
+
+// Stop stops the event processing and unsubscribes.
+func (b *BlockchainHealthManager) Stop() {
+	b.mu.Lock()
+
+	for _, cancel := range b.cancelFuncs {
+		cancel()
+	}
+	clear(b.cancelFuncs)
+	clear(b.providers)
+
+	b.mu.Unlock()
+	b.wg.Wait()
+}
+
+// Subscribe allows clients to receive notifications about changes.
+func (b *BlockchainHealthManager) Subscribe() chan struct{} {
+	ch := make(chan struct{}, 1)
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.subscribers = append(b.subscribers, ch)
+	return ch
+}
+
+// Unsubscribe removes a subscriber from receiving notifications.
+func (b *BlockchainHealthManager) Unsubscribe(ch chan struct{}) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// Remove the subscriber channel from the list
+	for i, subscriber := range b.subscribers {
+		if subscriber == ch {
+			b.subscribers = append(b.subscribers[:i], b.subscribers[i+1:]...)
+			close(ch)
+			break
+		}
+	}
+}
+
+// aggregateAndUpdateStatus collects statuses from all providers and updates the overall and short status.
+func (b *BlockchainHealthManager) aggregateAndUpdateStatus(ctx context.Context) {
+	b.mu.Lock()
+
+	// Collect statuses from all providers
+	providerStatuses := make([]rpcstatus.ProviderStatus, 0)
+	for _, provider := range b.providers {
+		providerStatuses = append(providerStatuses, provider.Status())
+	}
+
+	// Update the aggregator with the new list of provider statuses
+	b.aggregator.UpdateBatch(providerStatuses)
+
+	// Get the new aggregated full and short status
+	newShortStatus := b.getStatusPerChain()
+	b.mu.Unlock()
+
+	// Compare full and short statuses and emit if changed
+	if b.lastStatus == nil || !compareShortStatus(newShortStatus, *b.lastStatus) {
```

Comment:
hmm if writing `b.lastStatus` is protected by a mutex, shouldn't the reading also be?

Reply:
True, fixed

Reply:
fixed